packages/processors/src/strategy/donationVotingMerkleDistributionDirectTransfer/handlers/allocated.handler.ts

```diff
@@ -0,0 +1,198 @@
+import { Address, encodePacked, getAddress, keccak256 } from "viem";
+
+import { Application, Changeset, Donation, Round } from "@grants-stack-indexer/repository";
+import { ChainId, getToken, ProcessorEvent, Token } from "@grants-stack-indexer/shared";
+
+import { getTokenAmountInUsd, getUsdInTokenAmount } from "../../../helpers/index.js";
+import {
+    ApplicationNotFound,
+    IEventHandler,
+    ProcessorDependencies,
+    RoundNotFound,
+    UnknownToken,
+} from "../../../internal.js";
+import { ApplicationMetadata, ApplicationMetadataSchema } from "../../../schemas/index.js";
+
+type Dependencies = Pick<
+    ProcessorDependencies,
+    "roundRepository" | "applicationRepository" | "pricingProvider"
+>;
+
+export class OriginMissing extends Error {
+    constructor() {
+        super("Origin field is required");
+    }
+}
+
+export class MetadataParsingFailed extends Error {
+    constructor(additionalInfo?: string) {
+        super(`Failed to parse application metadata: ${additionalInfo}`);
+    }
+}
+
+/**
+ * Handles the Allocated event for the Donation Voting Merkle Distribution Direct Transfer strategy.
+ *
+ * This handler performs the following core actions when a donation is allocated to a project:
+ * - Validates that both the round and application exist
+ * - Retrieves token price data to calculate USD amounts
+ * - Creates a new donation record with the allocated amount
+ * - Links the donation to both the application and round
+ */
+export class DVMDAllocatedHandler implements IEventHandler<"Strategy", "AllocatedWithOrigin"> {
+    constructor(
+        readonly event: ProcessorEvent<"Strategy", "AllocatedWithOrigin">,
+        private readonly chainId: ChainId,
+        private readonly dependencies: Dependencies,
+    ) {}
+
+    /**
+     * Handles the AllocatedWithOrigin event for the Donation Voting Merkle Distribution Direct Transfer strategy.
+     * @returns {Changeset[]} The changeset containing an InsertDonation change
+     * @throws {OriginMissing} if the origin is missing
+     * @throws {RoundNotFound} if the round does not exist
+     * @throws {ApplicationNotFound} if the application does not exist
+     * @throws {UnknownToken} if the token does not exist
+     * @throws {TokenPriceNotFoundError} if the token price is not found
+     * @throws {MetadataParsingFailed} if the metadata is invalid
+     */
+    async handle(): Promise<Changeset[]> {
+        const { srcAddress } = this.event;
+        const { recipientId: _recipientId, amount, token: _token } = this.event.params;
+
+        if (!this.event.params.origin) {
+            throw new OriginMissing();
+        }
```

Comment:
i have some doubts regarding this.  Can origin be undefined or null ? Since it is an address the only way to represent null is with `0x000...`.

Another thing that is important to mention is that adding an extra layer of runtime validation (for example using zod) for events would be really helpful on these cases

Reply:
i had the same doubt but they have the validation in place, guided by the contracts, origin is always defined

Reply:
Maybe what they were handling all `Allocated` events as one and needed to validate it. In that case make sense to do it.

Reply:
that makes sense 🫡